rasa/graph_components/validators/finetuning_validator.py

```diff
@@ -0,0 +1,341 @@
+                f"training graph. "
+            )
+
+        # check the rasa version
+
+        if finetuning:
+            old_rasa_version = self._fingerprints[FINGERPRINT_VERSION]
+            if version.parse(old_rasa_version) < version.parse(
+                MINIMUM_COMPATIBLE_VERSION
+            ):
+                raise InvalidConfigException(
+                    f"The minimum compatible Rasa Version is "
+                    f"{MINIMUM_COMPATIBLE_VERSION} but the model we attempt to "
+                    f"finetune has been generated with an older version "
+                    f"({old_rasa_version}."
+                )
+        self._fingerprints[FINGERPRINT_VERSION] = rasa.__version__
+
+        # check the configuration
+
+        self._compare_raise_update(
+            fingerprint_key=FINGERPRINT_CONFIG_WITHOUT_EPOCHS_KEY,
+            new_fingerprint=self._get_fingerprint_of_config_without_epochs(config),
+            error_message=(
+                "Cannot finetune after making certain changes to the config. "
+                "The only changes that are allowed are"
+            ),
+        )
+
+        # check "core", i.e. domain
+
+        if core:
+            domain = importer.get_domain()
+            self._compare_raise_update(
+                fingerprint_key=FINGERPRINT_CORE,
+                new_fingerprint=self._get_fingerprint_of_domain_without_responses(
+                    domain
+                ),
+                error_message=(
+                    "Cannot finetune after making changes to the domain (only "
+                    "the responses may differ)."
+                ),
+            )
+
+        # check "nlu", i.e. nlu training data
+        # Note: Since this involves importing nlu data, it is probably the most
+        # costly step and hence happens last.
+
+        # NOTE: If there's a consistency check between domain and training data
+        # that ensures domain and NLU training data are consistent, then we can
+        # drop this check.
+
+        if nlu:
+            nlu_data = importer.get_nlu_data()
+            domain = importer.get_domain()
+            self._compare_raise_update(
+                fingerprint_key=FINGERPRINT_NLU,
+                new_fingerprint=nlu_data.label_fingerprint(),
+                error_message=(
+                    "Cannot finetune after inserting new NLU labels or removing "
+                    "NLU labels. Please make sure that the NLU data that you use "
+                    "for finetuning contains at least one example for every label "
+                    "(i.e. intent, action name, ...) that was included in the NLU "
+                    "data used for training the model which we attempt to finetune "
+                    "now. Moreover, you must not add labels that were not included "
+                    "during training before. "
+                ),
+            )
+
+        # persist
+        if not finetuning:
```

Comment:
What if we want to finetune a second time?

Reply:
then we can use the same fingerprints again - they haven't changed the first time we finetuned

Reply:
Actually the version could change. Although, if the updated version is compatible with the last version but incompatible with the oldest version, that would be strange. I can still update it...